### ISSUE TITLE
Bug: Parser returning a falsy value not handled well by JS language

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [eslint/eslint#19440]
### Environment

Node version: v22.12.0
npm version: v10.9.0
Local ESLint version: v9.20.1 (Currently used)
Global ESLint version: Not found
Operating System: darwin 22.3.0

### What parser are you using?

Other

### What did you do?

<details>
<summary>Configuration</summary>

```js
export default [
  {
    languageOptions: {
      parser: {
        parse: (...args) => {
          console.log("Parsing with:", { args });
        },
      },
    },
  },
];
```
</details>

```js
// n/a
```

### What did you expect to happen?

An informative error, such as:

```plaintext
Oops! Something went wrong! :(

ESLint: 9.20.1

Error: parse returned undefined instead of an AST object.
    at validate (/Users/josh/repos/repros/node_modules/eslint/lib/languages/js/source-code/source-code.js:62:15)
    at new SourceCode (/Users/josh/repos/repros/node_modules/eslint/lib/languages/js/source-code/source-code.js:365:9)
    ...
```

### What actually happened?

```plaintext
Oops! Something went wrong! :(

ESLint: 9.20.1

TypeError: Cannot read properties of undefined (reading 'tokens')
    at validate (/Users/josh/repos/repros/node_modules/eslint/lib/languages/js/source-code/source-code.js:61:14)
    at new SourceCode (/Users/josh/repos/repros/node_modules/eslint/lib/languages/js/source-code/source-code.js:365:9)
    ...
```

### Link to Minimal Reproducible Example

https://github.com/JoshuaKGoldberg/repros/tree/eslint-flat-config-empty-parse

### Participation

- [x] I am willing to submit a pull request for this issue.

### Additional comments

Other invalid ASTs are handled well by the JS `SourceCode` `validate()` method. For example, an empty `{}` produces:

```plaintext
Oops! Something went wrong! :(

ESLint: 9.20.1

Error: AST is missing the tokens array.
    at validate (/Users/josh/repos/repros/node_modules/eslint/lib/languages/js/source-code/source-code.js:62:15)
    at new SourceCode (/Users/josh/repos/repros/node_modules/eslint/lib/languages/js/source-code/source-code.js:365:9)
    ...
```

**Comment by nzakas** [comment]
Sounds good. More informative error messages are always a good thing.

**Comment by fasttime** [comment]
I think this has been addressed in #19458. Is there any work left to do?
